CAIPs/caip-319.md

```diff
@@ -0,0 +1,112 @@
+---
+caip: 319
+title: Wallet Notification JSON-RPC Method
+author: Alex Donesky (@adonesky1)
+discussions-to: https://github.com/ChainAgnostic/CAIPs/pull/X
+status: Draft
+type: Standard
+created: 2024-08-01
+requires: 2, 25, 171, 217
+---
+
+## Simple Summary
+
+CAIP-319 defines a JSON-RPC method for a wallet to send notifications to a
+caller in a specified context defined by a valid [scopeObject][CAIP-217] and,
```

Comment:
```suggestion
caller in a specified context authorized by or in a valid [scopeObject][CAIP-217] and,
```

This is a slightly obtuse phrasing, but I think it captures what I mean about the case where CAIP-25 authorized
```json
{
...
  sessionScopes: {
    "bip122:1" { ... },
    "eip155": {
      "scopes": ["1","42069"],
      ...
    }
...
}
```

and a later call to CAIP-319 or CAIP-27 targets `eip155:1`.  the technicality is that `eip155:1` is NOT a valid scopeObject in this case, even though it's a valid target authorized IN a valid scopeObject.

Reply:
Yep. Personally, per my comment [here](https://github.com/ChainAgnostic/CAIPs/pull/319#issuecomment-2265503607) I would think we should change the wording in CAIP-27 to exclude the possibility of calling CAIP-27 (or CAIP-319) with anything other than a **fully formed CAIP-2 chainId**  for the `scope`

Reply:
https://github.com/ChainAgnostic/CAIPs/pull/319/commits/930ed61828f32ecb2eb23888db6cb9fd775f9735 shows what I mean. I think we should make this change in CAIP-27 as well

Reply:
Its inaccurate to characterize the CAIP-27 `scope` param as a `scopeObject`... https://github.com/ChainAgnostic/CAIPs/blob/6922e7446cd90add44feb899417164b8f0a7a1e2/CAIPs/caip-27.md?plain=1#L85

Its a CAIP-2 chainId?

Reply:
whoops, it currently should read "the [`scopeString`](https://chainagnostic.org/CAIPs/caip-217#definition) of a valid `scopeObject`, which _could_ be a CAIP-2 or a namespace, that's what was meant by the current text. and I agree we should make a normative change, I'm just kind of stuck on the best change to make:

1.) any "authorized" CAIP-2 is legal, even if it refers to one [of potentially many] `scopes[]` members _in_ a namespace-wide, "carbon copy" scopeObject.  This puts the onus on the wallet to make sure that CAIP-2 has been authorized, which is probably fine sense every wallet just thinks of a namespace-wide authorization with multiple scopes as separate per-chain authorization objects anyways under the hood.

or

2.) be more explicit and say that ONLY a `scopeString` that IS ALSO a CAIP-2 AND has been authorized explicitly, removing the possibility of [case #2 here](https://github.com/ChainAgnostic/CAIPs/pull/319#issuecomment-2265126837)

Reply:
> 2.) be more explicit and say that ONLY a scopeString that IS ALSO a CAIP-2 AND has been authorized explicitly, removing the possibility of https://github.com/ChainAgnostic/CAIPs/pull/319#issuecomment-2265126837

This doesn't make sense to me. 

->>>

> 1.) any "authorized" CAIP-2 is legal, even if it refers to one [of potentially many] scopes[] members in a namespace-wide, "carbon copy" scopeObject, i.e. https://github.com/ChainAgnostic/CAIPs/pull/319#issuecomment-2265126837. This puts the onus on the wallet to make sure that a given CAIP-2 has been authorized, which is probably fine since every wallet just thinks of a namespace-wide authorization with multiple scopes as separate per-chain authorization objects anyways under the hood.

Definitely this. I don't see any reason for wallets (or the protocol generically) to treat or think of 

> scopes[] members in a namespace-wide, "carbon copy" scopeObject,

As different from individually enumerated scopes in any way at all. 

And then furthermore, per your comment [here](https://github.com/ChainAgnostic/CAIPs/pull/320#discussion_r1711286113), is there any outstanding reason we can't just constrain CAIP-27 (and this CAIP) to be only **full CAIP-2 chainIds** not `scopeStrings`?

Reply:
Fair enough! I'll open that PR before today's meeting.